### PR TITLE
Added Heavy Stander and Ping Types

### PR DESCRIPTION
**Pull request by xeroplz**
- Heavy Stander (Active) skill added.
- Combat Action updated to support ping.
- Ping types added for future skill handlers. (Mana Deflector and
Natrual Shield)

### CODE REVIEW

File: src/ChannelServer/Skills/Combat/CombatMastery.cs [hunk 1]
```diff
@@ -103,6 +103,9 @@ public CombatSkillResult Use(Creature attacker, Skill skill, long targetEntityId
 				// Defense
 				Defense.Handle(aAction, tAction, ref damage);
 
+				// Heavy Stander
+				HeavyStander.Handle(target, aAction, tAction, ref damage);
```
Comment (exectails) [inline src/ChannelServer/Skills/Combat/CombatMastery.cs]: I don't remember, what happens if you dual wield attack an enemy with Heavy Stander and it pings, does it act like Defense? If so you would have to handle that.

You also have to handle Stability and Stun. You currently set Stun in the handler, but this is overriden for any attacks that aren't defended.

File: src/ChannelServer/Skills/Combat/HeavyStander.cs [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Aura.Channel.Skills.Combat
+{
+	/// <summary>
+	/// Handle for the Heavy Stander skill. (Active)
```
Comment (exectails) [inline src/ChannelServer/Skills/Combat/HeavyStander.cs]: "Active"? Which Heavy Stander is "active"? o.o And if actually is active, does that mean mobs have to activate it first to use it? That sounds wrong.

File: src/ChannelServer/Skills/Combat/HeavyStander.cs [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+		/// Stun time for attacker. (Milliseconds)
+		/// </summary>
+		private const int attackerStun = 1000;
+
+		/// <summary>
+		/// Subscribes the handler to events required for training.
+		/// </summary>
+		public void Init()
+		{
+			// Empty due to unknown training methods.
```
Comment (exectails) [inline src/ChannelServer/Skills/Combat/HeavyStander.cs]: From what I can tell you can't train it, you either *have* rX or you *don't*.

File: src/ChannelServer/Skills/Combat/HeavyStander.cs [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+		}
+
+		public override StartStopResult Start(Creature creature, Skill skill, Mabi.MabiDictionary dict)
+		{
+			skill.State = SkillState.Ready;
```
Comment (exectails) [inline src/ChannelServer/Skills/Combat/HeavyStander.cs]: Don't use Ready for this. We assume that skills are "Ready" if they were loaded via preparing the skill. This doesn't mean it *has* to cause problems, but I wouldn't want to risk it. Normally start/stop skills have a state or condition that they set, maybe take a look at those, see if you can find something that matches the skill. If not, make a new field in CreatureTemp or create a new SkillState, "Active", or "Enabled", or something like that.

File: src/ChannelServer/Skills/Combat/HeavyStander.cs [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+		/// <param name="aAction"></param>
+		/// <param name="tAction"></param>
+		/// <param name="damage"></param>
+		/// <returns></returns>
+		public static void Handle(Creature user, AttackerAction aAction, TargetAction tAction, ref float damage)
+		{
+			// Setup
+			var heavyStanderSkill = user.Skills.Get(SkillId.HeavyStander);
+			var heavyStanderLevel = 0;
+			var rnd = new Random();
```
Comment (exectails) [inline src/ChannelServer/Skills/Combat/HeavyStander.cs]: Use RandomProvider, and do it after the null check, so to not waste resources if the target doesn't have Heavy Stander.

File: src/ChannelServer/Skills/Combat/HeavyStander.cs [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+
+			if (heavyStanderSkill == null)
+				return;
+
+			// Check for Level 1 - Used in on-screen messages
+			if (user.Skills.Has(SkillId.HeavyStander, SkillRank.RF) || user.Skills.Has(SkillId.HeavyStander, SkillRank.RE) || user.Skills.Has(SkillId.HeavyStander, SkillRank.RD) || user.Skills.Has(SkillId.HeavyStander, SkillRank.RC) || user.Skills.Has(SkillId.HeavyStander, SkillRank.RB) || user.Skills.Has(SkillId.HeavyStander, SkillRank.RA))
```
Comment (exectails) [inline src/ChannelServer/Skills/Combat/HeavyStander.cs]: You already got the skill above, these checks get the skill again and again, over and over, which is a waste. Check heavyStanderSkill.Info.Rank instead.

File: src/ChannelServer/Skills/Combat/HeavyStander.cs [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+				{
+					if (heavyStanderLevel == 0)
+					{
+						// This shouldn't be possible, but checking again before any actions happen.
+						return;
+					}
+
+					// Update actions
+
+					// Target = User
```
Comment (exectails) [inline src/ChannelServer/Skills/Combat/HeavyStander.cs]: You should mention this in the doc comment for the parameter, so the one calling this method knows what he has to pass.

File: src/ChannelServer/Skills/Combat/HeavyStander.cs [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+					damage = Math.Max(1, damage - (damage * (heavyStanderSkill.RankData.Var1 / 100f)));
+
+					// On-Screen Messages
+					if (heavyStanderLevel == 1)
+					{
+						var messsageChance = rnd.Next(1, 5);
+						switch (messsageChance)
+						{
+							case 1:
+								Send.Notice(aAction.Creature, "My attack is being defended by a skill!");
```
Comment (exectails) [inline src/ChannelServer/Skills/Combat/HeavyStander.cs]: Don't forget `Localization.Get`.

File: src/ChannelServer/Skills/Combat/HeavyStander.cs [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+								Send.Notice(aAction.Creature, "The target takes no damage at all!");
+								break;
+							case 2:
+								Send.Notice(aAction.Creature, "This attack is completely useless!");
+								break;
+						}
+					}
+				}
+			}
+			else if (heavyStanderSkill.State == SkillState.None)
```
Comment (exectails) [inline src/ChannelServer/Skills/Combat/HeavyStander.cs]: By doing the checks first, and handling afterwards, you can get rid of several indentions, which makes the code easier to read and follow. Just a general tip.

```
// Check state
if (heavyStanderSkill.State == SkillState.None)
    return;

// Sanity check
if (heavyStanderLevel == 0)
    return;

// Check chance
if (rnd.NextDouble() > (heavyStanderSkill.RankData.Var3 / 100f))
    return;

// handle
```

File: src/ChannelServer/Skills/Combat/HeavyStander.cs (hunk 2 above)
Comment (xeroplz) [inline src/ChannelServer/Skills/Combat/HeavyStander.cs]: No, there's a Heavy Stander (Active) and a Heavy Stander (Passive). They're two different skills with the same function.

File: src/ChannelServer/Skills/Combat/HeavyStander.cs (hunk 8 above)
Comment (xeroplz) [inline src/ChannelServer/Skills/Combat/HeavyStander.cs]: I've never really understood Localization.Get :o
What purpose does it serve?

File: src/ChannelServer/Skills/Combat/CombatMastery.cs (hunk 1 above)
Comment (xeroplz) [inline src/ChannelServer/Skills/Combat/CombatMastery.cs]: If you dual wield, both hits have a chance at being auto-defended. If your first hit is auto-defended, you probably won't have time to land the second attack before the attacker hits you again. It doesn't really work like Defense, just changes stun times and damage.

File: src/ChannelServer/Skills/Combat/HeavyStander.cs (hunk 6 above)
Comment (xeroplz) [inline src/ChannelServer/Skills/Combat/HeavyStander.cs]: :pray: Needed to know how to use skill.Info.Rank checks, but didn't know what to put for parameters. I'll try again ; o;

File: src/ChannelServer/Skills/Combat/CombatMastery.cs (hunk 1 above)
Comment (xeroplz) [inline src/ChannelServer/Skills/Combat/CombatMastery.cs]: Just for reference: http://wiki.mabinogiworld.com/view/Auto_Defense
Stability stays the same, only stun time is reduced.

Comment (exectails) [inline src/ChannelServer/Skills/Combat/CombatMastery.cs]: Alright, but you still have to make it not override the stun time.

File: src/ChannelServer/Skills/Combat/HeavyStander.cs [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+			// Update actions
+
+			// Target = User
+			tAction.Type = CombatActionType.TakeHit;
+			tAction.PDef = PingType.Melee;
+			tAction.Stun = targetStun;
+			aAction.Stun = attackerStun;
+
+			// Reduce damage
+			#region DamageCalculations
```
Comment (exectails) [inline src/ChannelServer/Skills/Combat/HeavyStander.cs]: Please refrain from using regions, especially for small stuff like this, as specified in our coding conventions.

**Comment by exectails** [comment]
I'm still not entirely sure what this active Heavy Stander is. It looks like only newer monsters have it (G12+), and since it's *active*, it makes me think it's something special, meaning it's probably not on all the time. Why else would it be active? This could also mean that it works differently from what we expect it to do.

Anyway, we will need a general Heavy Stander helper, maybe in SkillHelper, though I was kinda hoping we wouldn't have to use that static class too often, to check all possible Heavy Standers. The way you use it right now you would have to call like half a dozen different handlers, for different standers, in every skill. And multiple standers would reduce damage and stun multiple times.

An alternative method, which might even be better, would be to give creatures a Heavy Stander chance property that is checked in a general helper. That property would calculate a chance, based on passive and enabled skills, transformations, equipment, etc. You would only have one handling method and one method for the chance logic. This would save you from calling multiple handlers and checking if they were successful or not.

This would only work if the Heavy Standers are all "the same" in functionality of course.

File: src/ChannelServer/Skills/Combat/CombatMastery.cs (hunk 1 above)
Comment (xeroplz) [inline src/ChannelServer/Skills/Combat/CombatMastery.cs]: Hmm? It has to in order to function properly o-o
Attacker Stun time is set to 1000, while Target Stun time is reduced to 0.

Comment (exectails) [inline src/ChannelServer/Skills/Combat/CombatMastery.cs]: >You currently set Stun in the handler, but this is overriden for any attacks that aren't defended.

Check the actual skill handlers, the values are set *again*, *overriding* whatever you've set.

**Comment by exectails** [comment]
Closing for now, no progress in almost 2 months.

**Pull request closed without merging**